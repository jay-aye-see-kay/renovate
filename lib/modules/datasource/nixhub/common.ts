export const defaultRegistryUrl = 'https://www.nixhub.io/';

export const datasource = 'nixhub';

export const nixhubPackageMap = {
  nodejs: 'node',
};
