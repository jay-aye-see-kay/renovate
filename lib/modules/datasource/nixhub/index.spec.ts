import { getPkgReleases } from '..';
import { Fixtures } from '../../../../test/fixtures';
import * as httpMock from '../../../../test/http-mock';
import { EXTERNAL_HOST_ERROR } from '../../../constants/error-messages';
import { datasource, defaultRegistryUrl } from './common';

const packageName = 'nodejs';

function getPath(packageName: string): string {
  return `/packages/${packageName}?_data=routes/_nixhub.packages.$pkg._index`;
}

describe('modules/datasource/nixhub/index', () => {
  describe('getReleases', () => {
    it('throws for error', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .get(getPath(packageName))
        .replyWithError('error');
      await expect(
        getPkgReleases({
          datasource,
          packageName,
        }),
      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    });
  });

  it('returns null for 404', async () => {
    httpMock.scope(defaultRegistryUrl).get(getPath(packageName)).reply(404);
    expect(
      await getPkgReleases({
        datasource,
        packageName,
      }),
    ).toBeNull();
  });

  it('returns null for empty result', async () => {
    httpMock.scope(defaultRegistryUrl).get(getPath(packageName)).reply(200, {});
    expect(
      await getPkgReleases({
        datasource,
        packageName,
      }),
    ).toBeNull();
  });

  it('returns null for empty 200 OK', async () => {
    httpMock
      .scope(defaultRegistryUrl)
      .get(getPath(packageName))
      .reply(200, { versions: [] });
    expect(
      await getPkgReleases({
        datasource,
        packageName,
      }),
    ).toBeNull();
  });

  it('throws for 5xx', async () => {
    httpMock.scope(defaultRegistryUrl).get(getPath(packageName)).reply(502);
    await expect(
      getPkgReleases({
        datasource,
        packageName,
      }),
    ).rejects.toThrow(EXTERNAL_HOST_ERROR);
  });

  it('processes real data', async () => {
    httpMock
      .scope(defaultRegistryUrl)
      .get(getPath(packageName))
      .reply(200, Fixtures.get('releases.json'));
    const res = await getPkgReleases({
      datasource,
      packageName,
    });
    expect(res).toMatchSnapshot();
    expect(res?.releases).toHaveLength(19);
  });
});
