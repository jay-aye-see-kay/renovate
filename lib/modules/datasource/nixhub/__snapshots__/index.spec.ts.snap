// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/datasource/nixhub/index processes real data 1`] = `
{
  "homepage": "https://www.nixhub.io",
  "registryUrl": "https://www.nixhub.io",
  "releases": [
    {
      "version": "20.10.0",
    },
    {
      "version": "20.11.0",
    },
    {
      "version": "20.11.1",
    },
    {
      "version": "20.12.2",
    },
    {
      "version": "21.0.0",
    },
    {
      "version": "21.1.0",
    },
    {
      "version": "21.2.0",
    },
    {
      "version": "21.3.0",
    },
    {
      "version": "21.4.0",
    },
    {
      "version": "21.5.0",
    },
    {
      "version": "21.6.0",
    },
    {
      "version": "21.6.1",
    },
    {
      "version": "21.6.2",
    },
    {
      "version": "21.7.0",
    },
    {
      "version": "21.7.1",
    },
    {
      "version": "21.7.2",
    },
    {
      "version": "21.7.3",
    },
    {
      "version": "22.0.0",
    },
    {
      "version": "22.2.0",
    },
  ],
}
`;
